bootloader/lk.go

```diff
@@ -49,49 +74,228 @@ func newLk(rootdir string, opts *Options) Bootloader {
 		//      are very different from runtime vs image-building mode.
 		//
 		// determine mode we are in, runtime or image build
-		l.inRuntimeMode = !opts.PrepareImageTime
-	}
 
-	return l
+		l.prepareImageTime = opts.PrepareImageTime
+
+		l.role = opts.Role
+	}
 }
 
 func (l *lk) Name() string {
 	return "lk"
 }
 
 func (l *lk) dir() string {
-	// we have two scenarios, image building and runtime
-	// during image building we store environment into file
-	// at runtime environment is written directly into dedicated partition
-	if l.inRuntimeMode {
+	if l.prepareImageTime {
+		// at prepare-image time, then use rootdir and look for /boot/lk/ -
+		// this is only used in prepare-image time where the binary files exist
+		// extracted from the gadget
+		return filepath.Join(l.rootdir, "/boot/lk/")
+	}
+
+	// for runtime, we should only be using dir() for V1 and the dir is just
+	// the udev by-partlabel directory
+	switch l.role {
+	case RoleSole:
+		// TODO: this should be adjusted to try and use the kernel cmdline
+		//       parameter for the disk that the bootloader says to find
+		//       the lk partitions on if provided like the UC20 case does, but
+		//       that involves changing many more tests, so let's do that in a
+		//       followup PR
 		return filepath.Join(l.rootdir, "/dev/disk/by-partlabel/")
+	case RoleRecovery, RoleRunMode:
+		// TODO: maybe panic'ing here is a bit harsh...
+		panic("internal error: shouldn't be using lk.dir() for uc20 runtime modes!")
+	default:
+		panic("unexpected bootloader role for lk dir")
 	}
-	return filepath.Join(l.rootdir, "/boot/lk/")
 }
 
 func (l *lk) InstallBootConfig(gadgetDir string, opts *Options) error {
+	// make sure that the opts are put into the object
+	l.processOpts(opts)
 	gadgetFile := filepath.Join(gadgetDir, l.Name()+".conf")
-	systemFile := l.envFile()
+	// since we are just installing static files from the gadget, there is no
+	// backup to copy, the backup will be created automatically (if allowed) by
+	// lkenv when we go to Save() the environment file.
+	systemFile, err := l.envBackstore(primaryStorage)
+	if err != nil {
+		return err
+	}
 	return genericInstallBootConfig(gadgetFile, systemFile)
 }
 
 func (l *lk) Present() (bool, error) {
-	return osutil.FileExists(l.envFile()), nil
+	// if we are in prepare-image mode or in V1, just check the env file
+	if l.prepareImageTime || l.role == RoleSole {
+		primary, err := l.envBackstore(primaryStorage)
+		if err != nil {
+			return false, err
+		}
+
+		if osutil.FileExists(primary) {
+			return true, nil
+		}
+
+		// if the primary backstore doesn't exist, check the backup storage
```

Comment:
looking at it implemented, this bit here is probably right only for RoleSole though, at prepareImageTime we don't expect to rely only on the backup I think

Reply:
this will be done in a follow-up

Reply:
done in https://github.com/snapcore/snapd/pull/9756